File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Storage/ReceiptStorage.swift
//
//  ReceiptStorage.swift
//  PaykitDemo
//
//  Persistent storage for receipts using Keychain.
//

import Foundation

/// Manages persistent storage of payment receipts
class ReceiptStorage {
    
    private let keychain: KeychainStorage
    private let identityName: String
    private let maxReceiptsToKeep = 500  // Limit stored receipts
    
    // In-memory cache
    private var receiptsCache: [Receipt]?
    
    private var receiptsKey: String {
        "paykit.receipts.\(identityName)"
    }
    
    init(identityName: String, keychain: KeychainStorage = KeychainStorage(serviceIdentifier: "com.paykit.demo")) {
        self.identityName = identityName
        self.keychain = keychain
    }
    
    // MARK: - CRUD Operations
    
    /// Get all receipts (newest first)
    func listReceipts() -> [Receipt] {
        if let cached = receiptsCache {
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: receiptsKey) else {
                return []
            }
            var receipts = try JSONDecoder().decode([Receipt].self, from: data)
            // Sort by date, newest first
            receipts.sort { $0.createdAt > $1.createdAt }
            receiptsCache = receipts
            return receipts
        } catch {
            print("ReceiptStorage: Failed to load receipts: \(error)")
            return []
        }
    }
    
    /// Get receipts filtered by status
    func listReceipts(status: PaymentStatus) -> [Receipt] {
        return listReceipts().filter { $0.status == status }
    }
    
    /// Get receipts filtered by direction
    func listReceipts(direction: PaymentDirection) -> [Receipt] {
        return listReceipts().filter { $0.direction == direction }
    }
    
    /// Get recent receipts (limited count)
    func recentReceipts(limit: Int = 10) -> [Receipt] {
        return Array(listReceipts().prefix(limit))
    }
    
    /// Get a specific receipt
    func getReceipt(id: String) -> Receipt? {
        return listReceipts().first { $0.id == id }
    }
    
    /// Add a new receipt
    func addReceipt(_ receipt: Receipt) throws {
        var receipts = listReceipts()
        
        // Add new receipt at the beginning (newest first)
        receipts.insert(receipt, at: 0)
        
        // Trim to max size
        if receipts.count > maxReceiptsToKeep {
            receipts = Array(receipts.prefix(maxReceiptsToKeep))
        }
        
        try persistReceipts(receipts)
    }
    
    /// Update an existing receipt
    func updateReceipt(_ receipt: Receipt) throws {
        var receipts = listReceipts()
        
        guard let index = receipts.firstIndex(where: { $0.id == receipt.id }) else {
            throw ReceiptStorageError.notFound(id: receipt.id)
        }
        
        receipts[index] = receipt
        try persistReceipts(receipts)
    }
    
    /// Delete a receipt
    func deleteReceipt(id: String) throws {
        var receipts = listReceipts()
        receipts.removeAll { $0.id == id }
        try persistReceipts(receipts)
    }
    
    /// Search receipts by counterparty or memo
    func searchReceipts(query: String) -> [Receipt] {
        let query = query.lowercased()
        return listReceipts().filter { receipt in
            receipt.displayName.lowercased().contains(query) ||
            receipt.counterpartyKey.lowercased().contains(query) ||
            (receipt.memo?.lowercased().contains(query) ?? false)
        }
    }
    
    /// Get receipts for a specific counterparty
    func receiptsForCounterparty(publicKey: String) -> [Receipt] {
        return listReceipts().filter { $0.counterpartyKey == publicKey }
    }
    
    /// Clear all receipts
    func clearAll() throws {
        try persistReceipts([])
    }
    
    // MARK: - Statistics
    
    /// Total sent amount
    func totalSent() -> UInt64 {
        return listReceipts(direction: .sent)
            .filter { $0.status == .completed }
            .reduce(0) { $0 + $1.amountSats }
    }
    
    /// Total received amount
    func totalReceived() -> UInt64 {
        return listReceipts(direction: .received)
            .filter { $0.status == .completed }
            .reduce(0) { $0 + $1.amountSats }
    }
    
    /// Count of completed transactions
    func completedCount() -> Int {
        return listReceipts(status: .completed).count
    }
    
    /// Count of pending transactions
    func pendingCount() -> Int {
        return listReceipts(status: .pending).count
    }
    
    // MARK: - Private
    
    private func persistReceipts(_ receipts: [Receipt]) throws {
        let data = try JSONEncoder().encode(receipts)
        try keychain.store(key: receiptsKey, data: data)
        receiptsCache = receipts
    }
}

// MARK: - Errors

enum ReceiptStorageError: LocalizedError {
    case encodingFailed
    case decodingFailed
    case notFound(id: String)
    
    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode receipts"
        case .decodingFailed:
            return "Failed to decode receipts"
        case .notFound(let id):
            return "Receipt not found: \(id)"
        }
    }
}

File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/storage/ReceiptStorage.kt
package com.paykit.demo.storage

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.paykit.demo.model.PaymentDirection
import com.paykit.demo.model.PaymentStatus
import com.paykit.demo.model.Receipt
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/**
 * Manages persistent storage of payment receipts using EncryptedSharedPreferences.
 */
class ReceiptStorage(context: Context) {
    
    companion object {
        private const val TAG = "ReceiptStorage"
        private const val PREFS_NAME = "paykit_receipts"
        private const val RECEIPTS_KEY = "receipts_list"
        private const val MAX_RECEIPTS_TO_KEEP = 500
    }
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        encodeDefaults = true
    }
    
    private val prefs by lazy {
        try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            
            EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create encrypted prefs, falling back to regular prefs", e)
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }
    
    // In-memory cache
    private var receiptsCache: List<Receipt>? = null
    
    // MARK: - CRUD Operations
    
    /**
     * Get all receipts (newest first)
     */
    fun listReceipts(): List<Receipt> {
        receiptsCache?.let { return it }
        
        return try {
            val jsonString = prefs.getString(RECEIPTS_KEY, null) ?: return emptyList()
            val receipts = json.decodeFromString<List<Receipt>>(jsonString)
                .sortedByDescending { it.createdAt }
            receiptsCache = receipts
            receipts
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load receipts: ${e.message}")
            emptyList()
        }
    }
    
    /**
     * Get receipts filtered by status
     */
    fun listReceipts(status: PaymentStatus): List<Receipt> {
        return listReceipts().filter { it.status == status }
    }
    
    /**
     * Get receipts filtered by direction
     */
    fun listReceipts(direction: PaymentDirection): List<Receipt> {
        return listReceipts().filter { it.direction == direction }
    }
    
    /**
     * Get recent receipts (limited count)
     */
    fun recentReceipts(limit: Int = 10): List<Receipt> {
        return listReceipts().take(limit)
    }
    
    /**
     * Get a specific receipt
     */
    fun getReceipt(id: String): Receipt? {
        return listReceipts().find { it.id == id }
    }
    
    /**
     * Add a new receipt
     */
    fun addReceipt(receipt: Receipt) {
        val receipts = listReceipts().toMutableList()
        
        // Add new receipt at the beginning (newest first)
        receipts.add(0, receipt)
        
        // Trim to max size
        val trimmed = if (receipts.size > MAX_RECEIPTS_TO_KEEP) {
            receipts.take(MAX_RECEIPTS_TO_KEEP)
        } else {
            receipts
        }
        
        persistReceipts(trimmed)
    }
    
    /**
     * Update an existing receipt
     */
    fun updateReceipt(receipt: Receipt) {
        val receipts = listReceipts().toMutableList()
        val index = receipts.indexOfFirst { it.id == receipt.id }
        
        if (index >= 0) {
            receipts[index] = receipt
            persistReceipts(receipts)
        }
    }
    
    /**
     * Delete a receipt
     */
    fun deleteReceipt(id: String) {
        val receipts = listReceipts().toMutableList()
        receipts.removeAll { it.id == id }
        persistReceipts(receipts)
    }
    
    /**
     * Search receipts by counterparty or memo
     */
    fun searchReceipts(query: String): List<Receipt> {
        val lowerQuery = query.lowercase()
        return listReceipts().filter { receipt ->
            receipt.displayName.lowercase().contains(lowerQuery) ||
            receipt.counterpartyKey.lowercase().contains(lowerQuery) ||
            (receipt.memo?.lowercase()?.contains(lowerQuery) ?: false)
        }
    }
    
    /**
     * Get receipts for a specific counterparty
     */
    fun receiptsForCounterparty(publicKey: String): List<Receipt> {
        return listReceipts().filter { it.counterpartyKey == publicKey }
    }
    
    /**
     * Clear all receipts
     */
    fun clearAll() {
        persistReceipts(emptyList())
    }
    
    // MARK: - Statistics
    
    /**
     * Total sent amount
     */
    fun totalSent(): Long {
        return listReceipts(PaymentDirection.SENT)
            .filter { it.status == PaymentStatus.COMPLETED }
            .sumOf { it.amountSats }
    }
    
    /**
     * Total received amount
     */
    fun totalReceived(): Long {
        return listReceipts(PaymentDirection.RECEIVED)
            .filter { it.status == PaymentStatus.COMPLETED }
            .sumOf { it.amountSats }
    }
    
    /**
     * Count of completed transactions
     */
    fun completedCount(): Int {
        return listReceipts(PaymentStatus.COMPLETED).size
    }
    
    /**
     * Count of pending transactions
     */
    fun pendingCount(): Int {
        return listReceipts(PaymentStatus.PENDING).size
    }
    
    // MARK: - Private
    
    private fun persistReceipts(receipts: List<Receipt>) {
        try {
            val jsonString = json.encodeToString(receipts)
            prefs.edit().putString(RECEIPTS_KEY, jsonString).apply()
            receiptsCache = receipts
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save receipts: ${e.message}")
        }
    }
}

